Add iceThicknessHydro and waterPressureSmooth

alexolinhager opened the pull request:
This PR introduces two smoothed fields that help keep the hydro model stable when run with spatially variable bed topography and ice thickness. The first is iceThicknessHydro, which is a modified version of thickness that is only used by the hydrology model in place of the original thickness field. This PR moves through multiple iterations of iceThicknessHydro; however, the ultimate version is nearly identical to thickness but with all local minima and maxima removed. Local minima/maxima are defined as a single cell that has a thinner/thicker thickness than all of its immediate neighbors. Local minima/maxima are then removed by setting iceThicknessHydro at that cell equal to the mean of its neighbors. iceThicknessHydro is necessary to avoid channel and diffusivity instabilities that arise from single-cell minima and maxima. A new output variable, upperSurfaceHydro, is then introduced as the sum of bedTopography and iceThicknessHydro. The use of iceThicknessHydro can be toggled on and off with namelist option config_SGH_use_iceThicknessHydro.

The second change is the introduction of waterPressureSmooth, which is a smoothed version of waterPressure that is only used for calculation of waterPressureSlopeNormal. This updated version of waterPressureSlopeNormal has more gradual gradients that help keep channelPressureFreeze stable. However, channelPressureFreeze is not yet fully functional due to additional necessary improvements to marine boundary conditions in realistic geometries. This will be addressed in the next pull request.

matthewhoffman commented on components/mpas-albany-landice/src/Registry_subglacial_hydro.xml:
```diff
@@ -230,7 +246,9 @@
                 <var name="deltatSGH" type="real" dimensions="Time" units="s"
                      description="time step used for evolving subglacial hydrology system" />
-                <!-- channel variables -->
+					 <var name="pressureField" type="real" dimensions="nCells Time" units="Pa"
```
I have a note below to make this a local allocatable array and delete from Registry.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -860,7 +897,64 @@ subroutine calc_edge_quantities(block, err)
+      end do
+
+
+      ! Create a smoothed waterPressure product used only for
```
This subroutine is already pretty complicated and hard to follow.  Are you ok with moving the waterPressureSmooth calculation to a new subroutine that just gets called here?

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -860,7 +897,64 @@ subroutine calc_edge_quantities(block, err)
          hydropotentialBaseSlopeNormal(iEdge) = (hydropotentialBase(cell2) - hydropotentialBase(cell1)) / dcEdge(iEdge)
          hydropotentialSlopeNormal(iEdge) = (hydropotential(cell2) - hydropotential(cell1)) / dcEdge(iEdge)
-         waterPressureSlopeNormal(iEdge) = (waterPressure(cell2) - waterPressure(cell1)) / dcEdge(iEdge)
+         !waterPressureSlopeNormal(iEdge) = (waterPressure(cell2) - waterPressure(cell1)) / dcEdge(iEdge)
```
Can you call a water pressure smoothing subroutine before this edge loop and  leave this commented slope calculation here?

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -860,7 +897,64 @@ subroutine calc_edge_quantities(block, err)
+      ! calculating waterPressureSlopeNormal - to avoid channelPressureFreeze
+      ! instabilities
+
+      if (nTimesSmooth > 0) then
```
This if-statement can be removed - the do loop on the next line won't start if nTimeSmooth is 0.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -860,7 +897,64 @@ subroutine calc_edge_quantities(block, err)
+
+      if (nTimesSmooth > 0) then
+         do iter = 1, nTimesSmooth !May need to iterate to smooth properly
+            if (iter == 1) then      
```
Remove this if-statement and set these initial values before the do-loop.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -860,7 +897,64 @@ subroutine calc_edge_quantities(block, err)
+            endif
+
+            do iCell = 1, nCells
+               if ((li_mask_is_grounded_ice(cellMask(iCell))) .and. (.not. li_mask_is_margin(cellMask(iCell)))) then 
```
Can you add a comment explaining what the search criteria are?  It looks like you are excluding any locations that are along boundaries of the subglacial hydrology domain.  Does this include boundaries of a regional mesh (I forget if `li_mask_is_margin` is true on mesh boundaries)?

matthewhoffman replied:
Do you want to make use of your new li_mask_is_boundary function here to simplify this logic a bit?

alexolinhager replied:
Changing this to be all cells with grounded ice

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -860,7 +897,64 @@ subroutine calc_edge_quantities(block, err)
+                           endif
+                        end do
+                  
+                        !waterPressureSmooth is average of neighboring cells
```
```suggestion
                        !waterPressureSmooth is average of current cell and neighboring cells
```

matthewhoffman replied:
It would be better to make this an area-weighted average.  You need to accumulate both `areaCell* pressureField` and `areaCell` and then divide the two totals here.  I can dig up some examples if that would be useful.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -860,7 +897,64 @@ subroutine calc_edge_quantities(block, err)
+         cell1 = cellsOnEdge(1, iEdge)
+         cell2 = cellsOnEdge(2, iEdge)
+      
+         waterPressureSlopeNormal(iEdge) = (waterPressureSmooth(cell2) - waterPressureSmooth(cell1)) / dcEdge(iEdge)
```
`waterPressureSmooth` will need a halo update before using it because it's calculated based off of information about neighbors.

alexolinhager replied:
I don't think I've done this before, do you have an example of how to do this?

matthewhoffman replied:
The halo update itself is easy:
https://github.com/MALI-Dev/E3SM/blob/develop/components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F#L529-L531
but it needs to happen on the whole `domain` object, which is only available in the main SGH driver, so we might need to reorganize some things to be able to do that.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -860,7 +897,64 @@ subroutine calc_edge_quantities(block, err)
+         do iter = 1, nTimesSmooth !May need to iterate to smooth properly
+            if (iter == 1) then      
+               waterPressureSmooth = waterPressure
+               pressureField = waterPressure
```
Let's make pressureField an allocatable local array - no need to have it in Registry as a temporary variable.

alexolinhager replied:
Agreed, I tried to do this before but wasn't sure the best way to go about it

matthewhoffman replied:
Here's a straightforward example with an badly named variable :) 
https://github.com/MALI-Dev/E3SM/blob/8be18efd4a8bff2008d38388de7fa1d28709a4a8/components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F#L277
https://github.com/MALI-Dev/E3SM/blob/8be18efd4a8bff2008d38388de7fa1d28709a4a8/components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F#L450-L453

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2166,7 +2264,9 @@ subroutine ocean_connection_N(domain)
          call mpas_pool_get_array(geometryPool, 'thickness', thickness)
```
remove this line and the declaration for it above on line 2249

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -1665,14 +1761,15 @@ subroutine calc_pressure_diag_vars(block, err)
+      call mpas_pool_get_array(hydroPool, 'iceThicknessHydro', iceThicknessHydro)
       call mpas_pool_get_array(geometryPool, 'cellMask', cellMask)
 
-      effectivePressure = rhoi * gravity * thickness - waterPressure
```
remove the get_array call and variable declaration for `thickness` above now that we're no longer using it.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2270,5 +2371,160 @@ subroutine calc_hydro_mask(domain)
+      if (config_SGH_use_iceThicknessHydro) then
+         do iCell = 1, nCells
+
+            if ((li_mask_is_grounded_ice(cellMask(iCell))) .and. (.not. li_mask_is_margin(cellMask(iCell)))) then !identify domain boundaries
```
This PR includes changes to the mask routine to add li_mask_is_boundary.  Presumably you can replace this block of code with calling that instead?

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2270,5 +2371,160 @@ subroutine calc_hydro_mask(domain)
+               enddo
+
+               if (isMarineMargin == 0) then
+                  maxNeighborHeight = bigNegativeValue !allocate
```
```suggestion
                  maxNeighborHeight = bigNegativeValue ! initialize
```

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2270,5 +2371,160 @@ subroutine calc_hydro_mask(domain)
+
+                     if (iNeighbor < nCells + 1) then
+
+                         if (jCell == 1) then
+                                 totNeighborHeight = upperSurface(iNeighbor)
+                                 numCells = 1
+                         endif 
```
You can eliminate this if-statement by initializing `totNeighborHeight = 0.0_RKIND` and `numCells=0` before the do-loop over nEdgesOnCell.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2270,5 +2371,160 @@ subroutine calc_hydro_mask(domain)
+                     endif
+                  end do
+
+                  meanNeighborHeight = totNeighborHeight / numCells
```
As with waterPressureSmooth, would be better to do an area-weighted average

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2270,5 +2371,160 @@ subroutine calc_hydro_mask(domain)
+!> \author  Alex Hager
+!> \date    7 June 2023
+!> \details
+!>  This routine calculates a modified ice thickness that is altered at
+!   the domain boundaries to avoid local minima in hydropotential
```
It looks like the code is altering ice thickness everywhere *but* domain boundaries.  Is that the intention?  I remember originally this was implemented to deal with low spots along the domain boundary. Is this no longer doing that?

matthewhoffman commented on components/mpas-albany-landice/src/Registry_subglacial_hydro.xml:
```diff
@@ -77,7 +77,13 @@
-                <!-- channel options -->
+		
+		<nml_option name="config_SGH_use_iceThicknessHydro" type="logical" default_value=".true." units="unitless"
+			    description="alter ice thickness along boundaries to avoid local maxima/minima in upperSurface. This can circumvent instabilities from forming along the boundaries of the hydro model that arise from unrealistic surface anomalies left over from building the domain. This option has no significant effect on the behavior of the model but makes it more stable."
```
See comment below for the new subroutine - it looks like it currently is applied everywhere *except* domain boundaries.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -860,7 +897,64 @@ subroutine calc_edge_quantities(block, err)
+                        endif
+                     enddo
+
+                     if (isMarineMargin == 0) then 
```
better to make this type logical

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2270,5 +2371,160 @@ subroutine calc_hydro_mask(domain)
+                  endif
+               enddo
+
+               if (isMarineMargin == 0) then
```
better to make this type logical

matthewhoffman reviewed (changes requested):
@alexolinhager , thanks for putting together these improvements for the hydro model.  I have a number of code cleanup comments and a couple higher level questions to confirm about the intended behavior.  I'm happy to advise or help with revisions.

It would also be good to add example images of the thickness/iceThicknessHydro and waterPressure/waterPressureSmooth fields (and their differences) for a realistic mesh to do a visual confirmation they look as expected.  I'll also show you how to run the integration test suite, which will test for restart and decomposition issues.

alexolinhager commented:
@matthewhoffman, I finished going through your comments, and added three more commits that address your concerns. I believe I implemented the halo updates correctly, but it would be good for you to double check they're right. Does iceThicknessHydro also need a halo update if it is also calculated based on its neighbors? 

Also, below are a couple plots comparing `upperSurfaceHydro` (`upperSurfaceHydro` = `iceThicknessHydro` + `bedTopography`) with `upperSurface`, as well as `waterPressureSmooth` with `waterPressure`. Both are behaving as I would expect. Important to note that `iceThicknessHydro`/`upperSurfaceHydro' changes only anomalous cells, whereas `waterPressureSmooth` targets all  grounded ice cells.
![Screenshot 2024-03-04 at 9 33 45 AM](https://github.com/MALI-Dev/E3SM/assets/131483939/c37d6582-b0d9-4111-8700-702fd51e3565)
![Screenshot 2024-03-04 at 10 52 10 AM](https://github.com/MALI-Dev/E3SM/assets/131483939/ffdf8468-c30b-4952-af51-ee73c1a856ab)

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2464,41 +2465,37 @@ subroutine calc_iceThicknessHydro(block, err)
          do iCell = 1, nCells
 
-            if ((li_mask_is_grounded_ice(cellMask(iCell))) .and. (.not. li_mask_is_margin(cellMask(iCell)))) then !identify domain boundaries
+            if (li_mask_is_grounded_ice(cellMask(iCell))) then !identify domain boundaries
```
update comment

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2270,5 +2320,260 @@ subroutine calc_hydro_mask(domain)
+      call mpas_pool_get_array(meshPool, 'edgesOnCell', edgesOnCell)
+      call mpas_pool_get_array(meshPool, 'nEdgesOnCell', nEdgesOnCell)
+      call mpas_pool_get_array(meshPool, 'areaCell', areaCell)
+      call mpas_pool_get_array(hydroPool, 'hydroMarineMarginMask', hydroMarineMarginMask)
```
remove this line and the variable declaration above

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2270,5 +2320,260 @@ subroutine calc_hydro_mask(domain)
+      call mpas_pool_get_array(geometryPool, 'thickness', thickness)
+      call mpas_pool_get_array(geometryPool, 'bedTopography', bedTopography)
+      call mpas_pool_get_array(geometryPool, 'cellMask', cellMask)
+      call mpas_pool_get_array(geometryPool, 'edgeMask', edgeMask)
```
remove this line and the variable declaration above

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2270,5 +2320,260 @@ subroutine calc_hydro_mask(domain)
+      integer, dimension(:), pointer :: edgeMask
+      real (kind=RKIND), dimension(:), pointer :: areaCell
+      integer, pointer :: nCells
+      integer :: edgeNum
```
remove - not used

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2270,5 +2320,260 @@ subroutine calc_hydro_mask(domain)
+      integer, pointer :: nCells
+      integer :: edgeNum
+      integer, dimension(:), pointer :: nEdgesOnCell
+      integer :: iEdge
```
remove - not used

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2270,5 +2320,260 @@ subroutine calc_hydro_mask(domain)
+      call mpas_pool_get_array(meshPool, 'nEdgesOnCell', nEdgesOnCell)
+      call mpas_pool_get_dimension(meshPool, 'nCells', nCells)
+      call mpas_pool_get_array(meshPool, 'cellsOnCell', cellsOnCell)
+      call mpas_pool_get_array(meshPool, 'edgesOnCell', edgesOnCell)
```
remove line and variable declaration above - not used

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2270,5 +2320,260 @@ subroutine calc_hydro_mask(domain)
+          
+               waterPressureSmooth(iCell) = totalPressure / totalArea !area-weighted average pressure
+            endif
+      end do
+
+      pressureField = waterPressureSmooth
+
+   end do
+
+   deallocate(pressureField)
```
These lines should be indented to match the rest of the routine

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2270,5 +2320,260 @@ subroutine calc_hydro_mask(domain)
+               totalPressure = pressureField(iCell)*areaCell(iCell) 
+               totalArea = areaCell(iCell)
+          
+               do jCell = 1, nEdgesOnCell(iCell)
```
Minor detail - it would be more accurate to rename `jCell` to `jEdge`

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -195,9 +200,22 @@ subroutine li_SGH_init(domain, err)
          block => block % next
       end do
 
+      !update halo for iceThicknessHydro
+      call mpas_timer_start("halo updates")
+      call mpas_dmpar_field_halo_exch(domain, 'iceThicknessHydro')
+      call mpas_timer_stop("halo updates")
```
This needs to be moved up after iceThicknessHydro is calculated but before it is used inside the call to calc_pressure_diag_vars.  Otherwise the calculations done using it there will have bad values in the halos.  One detail I failed to mention is that halo updates should be outside of block loops. It doesn't really matter given we no longer support multiple blocks, but for consistency with existing code, let's continue to do that.  That will mean ending the block loop after calling `calc_iceThicknessHydro`, then calling the halo update, then starting a new block loop for the remaining steps.  Any pools or fields that are retrieved need to be retrieved again in the new block loop.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -611,6 +635,14 @@ subroutine li_SGH_solve(domain, err)
+
+         !updates halos for waterPressure
+         call mpas_timer_start("halo updates")
+         call mpas_dmpar_field_halo_exch(domain, 'waterPressureSlopeNormal')
```
Did you mean this halo update to be on waterPressureSmooth?

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -195,9 +200,22 @@ subroutine li_SGH_init(domain, err)
          call calc_pressure_diag_vars(block, err_tmp)
          err = ior(err, err_tmp)
 
+         call calc_waterPressureSmooth(block, err_tmp) !adjust ice thickness along boundaries
```
update the comment here

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -195,9 +200,22 @@ subroutine li_SGH_init(domain, err)
+
+         !updates halos for waterPressure
+         call mpas_timer_start("halo updates")
+         call mpas_dmpar_field_halo_exch(domain, 'waterPressureSlopeNormal')
```
Presumably you meant to do this on waterPressureSmooth?

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -783,19 +819,24 @@ subroutine calc_edge_quantities(block, err)
       integer, pointer :: nVertices
       integer :: iEdge, cell1, cell2
-      integer :: i, j, iVertex, iCell
+      integer :: i, j, iVertex, iCell, jCell
```
remove jCell - got moved to the new subroutine

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -783,19 +819,24 @@ subroutine calc_edge_quantities(block, err)
       character (len=StrKIND), pointer :: config_SGH_tangent_slope_calculation
       real (kind=RKIND), pointer :: config_sea_level
       real (kind=RKIND), pointer :: rhoo
+      integer, pointer :: nTimesSmooth
```
remove - got moved to the new subroutine

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -765,12 +797,16 @@ subroutine calc_edge_quantities(block, err)
       integer, dimension(:), pointer :: cellMask
       integer, dimension(:), pointer :: edgeMask
       integer, dimension(:,:), pointer :: cellsOnEdge
+      integer, dimension(:,:), pointer :: edgesOnCell
+      integer, dimension(:,:), pointer :: cellsOnCell
+      integer, dimension(:), pointer :: nEdgesOnCell
```
remove - got moved to new subroutine

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -783,19 +819,24 @@ subroutine calc_edge_quantities(block, err)
       integer :: numGroundedCells
       integer :: err_tmp
-
+      integer :: isMarineMargin
+      integer :: edgeNum
+      integer :: iNeighbor 
+      integer :: numCells
+      integer :: iter
```
remove - got moved to new subroutine

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -813,7 +854,8 @@ subroutine calc_edge_quantities(block, err)
       call mpas_pool_get_config(liConfigs, 'config_sea_level', config_sea_level)
       call mpas_pool_get_config(liConfigs, 'config_ocean_density', rhoo)
-
+      call mpas_pool_get_config(liConfigs, 'config_SGH_iter_smooth_waterPressureSlopeNormal', nTimesSmooth)
```
remove - got moved to new subroutine

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -828,6 +870,9 @@ subroutine calc_edge_quantities(block, err)
       call mpas_pool_get_array(meshPool, 'dcEdge', dcEdge)
       call mpas_pool_get_array(geometryPool, 'cellMask', cellMask)
       call mpas_pool_get_array(meshPool, 'cellsOnEdge', cellsOnEdge)
+      call mpas_pool_get_array(meshPool, 'edgesOnCell', edgesOnCell)
+      call mpas_pool_get_array(meshPool, 'cellsOnCell', cellsOnCell)
+      call mpas_pool_get_array(meshPool, 'nEdgesOnCell', nEdgesOnCell)
```
remove - got moved to new subroutine

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -842,8 +887,9 @@ subroutine calc_edge_quantities(block, err)
       call mpas_pool_get_array(geometryPool, 'edgeMask', edgeMask)
-
-
+      call mpas_pool_get_array(hydroPool, 'hydroMarineMarginMask', hydroMarineMarginMask)
```
remove - got moved to new subroutine

matthewhoffman reviewed (changes requested):
@alexolinhager , thanks for the updates on this PR.  I made a number of small cleanup requests, and then there are a few issues with the halo updates that need resolving.  Let me know if you have any questions about those details.

alexolinhager commented:
@matthewhoffman I went through the additional edits and will run the integration test suite to make sure everything still works before sending it back to you for review. Also, I'm not sure how the PR #78 got linked above but it was unintentional and I'm not sure how to undo it

alexolinhager commented:
As of now, this PR passes all tests except the baseline comparison for the four hydro test cases. Smoothing of the ice thickness and water pressure fields are expected to change results from baseline configuration so this is not a surprise
![Screenshot 2024-03-11 at 4 10 05 PM](https://github.com/MALI-Dev/E3SM/assets/131483939/f008c02d-2371-4bb4-b8fc-a0749af03afd)

alexolinhager commented:
Oddly, test validation fails in when waterPressureSmooth and iceThicknessHydro are turned off. 
![Screenshot 2024-03-11 at 4 09 08 PM](https://github.com/MALI-Dev/E3SM/assets/131483939/198e2c94-24cb-44bf-8478-5113869c58c5)

alexolinhager commented:
Also, perhaps unrelated, this lines missed editing after last review and need to be fixed: [https://github.com/MALI-Dev/E3SM/blob/5d139d93a96225fce40c0f3d99330855c09db7ec/com[…]mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F](https://github.com/MALI-Dev/E3SM/blob/5d139d93a96225fce40c0f3d99330855c09db7ec/components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F#L206-L209) and [https://github.com/MALI-Dev/E3SM/blob/5d139d93a96225fce40c0f3d99330855c09db7ec/com[…]mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F](https://github.com/MALI-Dev/E3SM/blob/5d139d93a96225fce40c0f3d99330855c09db7ec/components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F#L642C10-L645C46)

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2270,5 +2327,250 @@ subroutine calc_hydro_mask(domain)
+         enddo
+      endif
+
+      upperSurfaceHydro = iceThicknessHydro + bedTopography
```
Actually, it looks like the hydro module only uses upperSurface for the idealized SHMIP forcing.  We could also delete it if we don't need it.

alexolinhager replied:
I had kept `upperSurfaceHydro` because it was a helpful diagnostic to make sure `iceThicknessHydro` was doing what we wanted, but you're right that it's not needed anymore. I'll delete from here and the registry

matthewhoffman reviewed (changes requested):
@alexolinhager , I've done another round of review and have additional cleanup feedback.  We'll also need to get to the bottom of the restart test failures when iceThicknessHydro is turned off.  I don't see any obvious explanations.  If you look at the case_outputs log file, it will say which fields and time levels are failing validation, which should offer some clues.

alexolinhager commented:
@matthewhoffman, I rebased this branch onto `develop` and addressed your further comments above. The PR now passes the full integration test, so should be ready to merge if you're happy with it.
![Screenshot 2024-05-22 at 3 24 39 PM](https://github.com/MALI-Dev/E3SM/assets/131483939/84aea1f5-ca3f-4fe4-a1a4-cca0582158d4)

alexolinhager commented:
@matthewhoffman, I rebased this branch on top of `develop` and addressed your remaining comments above. All full integration hydro tests fail the baseline validation with iceThicknessHydro and waterPressureSmooth turned on, but that should be expected. I'm am about to rerun the suite with those fields turned off for an apples-to-apples comparison
![Screenshot 2024-05-22 at 3 50 27 PM](https://github.com/MALI-Dev/E3SM/assets/131483939/379c7860-66e2-4ebc-80ad-4c05522d56f9)

alexolinhager commented:
Shoot, looks like we're still getting the same problem as before with the restart runs when the smoothed variables are turned off. I'll need to dig into this more tomorrow. I don't understand what's causing these issues 
![Screenshot 2024-05-22 at 4 09 43 PM](https://github.com/MALI-Dev/E3SM/assets/131483939/bf93087c-ecb6-4a67-8ae9-a0400d8cce99)

matthewhoffman commented:
@alexolinhager , thanks for the update.  I'll try to look things over and confirm everything is ready, other than this restart issue.

alexolinhager commented:
@matthewhoffman, admittedly yesterday's problem was the result of end-of-the-day user error. I had only been changing configs in `namelist.landice` and had missed the `namelist.landice.rst` files in the restart runs. I made that change and all tests now pass with `iceThicknessHydro` and `waterPressureSmooth` turned off. The baseline comparison now only fails with those fields turned on, which is what we would expect. I think we're ready to merge if you're happy with it.
![Screenshot 2024-05-23 at 8 10 57 AM](https://github.com/MALI-Dev/E3SM/assets/131483939/702ea7f0-d80b-401d-ac60-17799b2c2d7b)
